cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -813,11 +813,17 @@ Status ParquetFileFragment::EnsureCompleteMetadata(parquet::arrow::FileReader* r
 
 Status ParquetFileFragment::SetMetadata(
     std::shared_ptr<parquet::FileMetaData> metadata,
-    std::shared_ptr<parquet::arrow::SchemaManifest> manifest) {
+    std::shared_ptr<parquet::arrow::SchemaManifest> manifest,
+    std::shared_ptr<parquet::FileMetaData> original_metadata) {
   DCHECK(row_groups_.has_value());
 
   metadata_ = std::move(metadata);
   manifest_ = std::move(manifest);
+  original_metadata_ = original_metadata ? std::move(original_metadata) : metadata_;
```

Comment:
This is ok for be but should we check `Subset` for this?

Reply:
Do you mean `FileMetaData::Subset`?

Reply:
Yeah, I misunderstand this previously, `metadata_` is from Subset, but `manifest` is shared and having lifetime problem here.